### PR TITLE
Closes #123 - Replace brctl with ip and bridge commands

### LINKED ISSUE #122: Last bridge disabling STP doesn't stop mstpd
If mstpd is running, and the last configured bridge has STP turned off, the bridge-stp removes the bridge, but doesn't kill the mstpd process.

The bug appears to be in the stop logic of bridge-stp, which checks if a bridge is in user-mode, but doesn't exclude the bridge being stopped (which is still in user-mode).

I can't supply a patch if desired...

### LINKED ISSUE #123: RFE: support distributions without brctl using iproute2
brctl isn't available on Centos 8 Stream (possibly others).

As brctl is critical to the functioning of the mstpd scripts (even bridge-stp when mstpd fails), it would be useful to have a fallback to iproute2 tools when it's not available:

 - brctl addbr $bridge ->  ip link add name $bridge type bridge
 - brctl delbr $bridge -> ip link del $bridge
 - brctl addif $bridge $port -> ip link set dev $port master $bridge
 - brctl delif $bridge $port -> ip link set dev $port nomaster
 - brctl stp $bridge {on|off} -> ip link set $bridge type bridge stp_state {state}

I found that all these worked even on OpenWRT with ip-tiny... perhaps it makes sense to just use ip for everything to reduce the dependencies?

I can work up a patch if that sounds right.

### PR DESCRIPTION
**Pull request by sshambar**
As brctl is deprecated in many distributions (and removed in
CentOS 8 Stream), this patch replaces it with the equivalent
ip/bridge commands from iproute2 (already required).

The patch also handles a race with spawning mstpd, corrects
killing it (Fixes #122), and has bridge-stp log errors to
syslog (if possible, as stderr is not logged).

Also fixes handling of pre-up etc in interfaces parsing.

Signed-off-by: Scott Shambarger <devel@shambarger.net>

### CODE REVIEW

File: bridge-stp.in [hunk 1]
```diff
@@ -3,7 +3,8 @@
 # Configuration is in @bridgestpconffile@
 #
 # `/sbin/bridge-stp <bridge> <start|stop>` is called by the kernel when STP is
-# enabled/disabled on a bridge (via `brctl stp <bridge> <on|off>`).  The kernel
+# enabled/disabled on a bridge (via
```
Comment (PaulSD) [inline bridge-stp.in:6]: In other places, you added `ip link` in addition to the `brctl`, instead of replacing `brctl` with `ip link` ... Would make sense to do that here too.

File: bridge-stp.in [hunk 2]
```diff
@@ -80,20 +81,29 @@ MSTPD_ARGS=''
 # A space-separated list of bridges for which MSTP should be used in place of
 # the kernel's STP implementation.  If empty, MSTP will be used for all bridges.
 MSTP_BRIDGES=''
+LOGGER='logger -t bridge-stp -s'
```
Comment (PaulSD) [inline bridge-stp.in:84]: Use --tag and --stderr so others can see what this is doing without having to check the man page for `logger`.

File: bridge-stp.in [hunk 3]
```diff
@@ -14,7 +15,7 @@
 # (possibly in a background job after a delay; see the comments in the code
 # below).  No further configuration is performed automatically by this script at
 # that time.  Additional configuration is usually performed by
-# @ifupdownfile@ (which calls `brctl stp <bridge> on`
+# @ifupdownfile@ (which calls `ip ... stp_state 1`
```
Comment (PaulSD) [inline bridge-stp.in:18]: Since you list it in full elsewhere, why not use `ip link set <bridge> type bridge stp_state 1` here too?

File: utils/ifupdown.sh.in [hunk 4]
```diff
@@ -93,15 +93,36 @@ done
 # We finish setting up the bridge
 if [ "$MODE" = 'start' ]; then
 
+  # check if mstpd is running before bridge-stp may spawn it
+  found=$(pidof -c -s mstpd)
```
Comment (PaulSD) [inline utils/ifupdown.sh.in:97]: Should probably be quoted: `found="$(...)"`.  Variable name could also be better ... maybe `mstpdpid`?

File: utils/ifupdown.sh.in [hunk 5]
```diff
@@ -93,15 +93,36 @@ done
 # We finish setting up the bridge
 if [ "$MODE" = 'start' ]; then
 
+  # check if mstpd is running before bridge-stp may spawn it
```
Comment (PaulSD) [inline utils/ifupdown.sh.in:96]: Capitalize "Check" to match other comments.

File: utils/ifupdown.sh.in [hunk 6]
```diff
@@ -93,15 +93,36 @@ done
...
+  found=$(pidof -c -s mstpd)
+
   # This triggers the kernel to run `/sbin/bridge-stp start $IFACE`
-  "$brctl" stp "$IFACE" on
+  "$ip" link set "$IFACE" type bridge stp_state 1
+
+  # if mstpd is just spawned, let it initialize
```
Comment (PaulSD) [inline utils/ifupdown.sh.in:102]: Capitalize "If" to match other comments.

File: utils/ifupdown.sh.in [hunk 7]
```diff
@@ -93,15 +93,36 @@ done
...
+  if [ -z "$found" ]; then
+    maxwait=2
+    # Use 0.1 delay if available
+    sleep 0.1 2>/dev/null && maxwait=$((maxwait*10))
+    count=0
+    while [ -z "$found" ] && [ $count -lt $maxwait ]; do
+      found=$(pidof -c -s mstpd)
```
Comment (PaulSD) [inline utils/ifupdown.sh.in:109]: I must be missing something ... Why is this necessary?  I'm pretty sure `brctl stp <iface> on` didn't return until after `/sbin/bridge-stp` returned, and `mstpd` was guaranteed to be running when `/sbin/bridge-stp` returned.  Does `ip link ... stp_state 1` behave differently and return early?

**Review by PaulSD — COMMENTED** [review]
Generally looks good, but I'd like to understand the spawning race condition better...

File: utils/ifupdown.sh.in (hunk 7 above)
Comment (sshambar) [reply utils/ifupdown.sh.in:109]: They behave the same (both wait), but there is a race which I can reproduce on both brctl and ip  link... just run:
```
# brctl stp <bridge> on; mstpctl addbridge <bridge>
ctl_client_init: Couldn't connect to server
can't setup control connection

```
Even the slightest delay (sleep 0.1) seems to overcome it, so I'm guessing mstpd backgrounds before it's ready to receive connections... this small loop is easy enough though :)

File: utils/ifupdown.sh.in (hunk 4 above)
Comment (sshambar) [reply utils/ifupdown.sh.in:97]: I'll update the variable name; quoting isn't necessary as word splitting doesn't apply here (eg. in bash or /bin/sh try x="a b"; y=$x; echo $y)... I could still add if you want (doesn't hurt), but then should probably be used in the $("$mstpctl" ...) further down in the same file...

Comment (PaulSD) [reply utils/ifupdown.sh.in]: Yeah, I'm aiming for consistency, and we quote stuff like that in several other places, so let's do it here too.  And yes, we should probably add quotes on curfwddly (that was added in another PR, and I must have missed it when reviewing that one).

File: utils/ifupdown.sh.in (hunk 7 above)
Comment (PaulSD) [reply utils/ifupdown.sh.in]: Hmm... Ok.  In that case, I think this loop/delay should probably go in the bridge-stp script (after spawning mstpd) instead of in ifupdown.sh.  bridge-stp could be used on systems that don't use ifupdown.sh (that use a different script for configuration), so let's solve this for everyone instead of only for those systems that use ifupdown.sh.

Comment (sshambar) [reply utils/ifupdown.sh.in]: OK, wasn't sure that would work given the warnings about kernel locks in the bridge-stp.in script, but I tried it and a brief sleep there allowed the client process to connect to mstpd, so I'll do that and revert the ifupdown loop.

File: utils/ifupdown.sh.in (hunk 4 above)
Comment (sshambar) [reply utils/ifupdown.sh.in]: OK, with the removal of the loop (sleep in bridge-stp), there's no pidof call anymore, but I'll quote the curfwddly anyway :)

**Comment by commodo** [comment]
@PaulSD 
any comments?

**Review by commodo — APPROVED** [review]

**Comment by commodo** [comment]
apologies for being too slow;
let's agree to a smaller timeout;

if nobody says anything for 2 weeks and i don't have any objections, i will merge it :)

**Comment by PaulSD** [comment]
I still want to make some tweaks here, but I know I'm taking forever to get around to this ... So go ahead and merge it and I'll clean it up later :)

**Comment by commodo** [comment]
> I still want to make some tweaks here, but I know I'm taking forever to get around to this ... So go ahead and merge it and I'll clean it up later :)

i'm fine with cleaning it up later;
